File: occasum-core/src/main/java/pw/stamina/occasum/node/visit/AbstractPropertyNodeVisitor.java
package pw.stamina.occasum.node.visit;

import pw.stamina.occasum.node.PropertyNode;

import java.util.Optional;

public abstract class AbstractPropertyNodeVisitor implements PropertyNodeVisitor {

    @Override
    public final Optional<PropertyNodeVisitor> visitChildNode(PropertyNode childNode) {
        return Optional.ofNullable(createChildVisitor(childNode));
    }

    protected abstract PropertyNodeVisitor createChildVisitor(PropertyNode childNode);
}

File: occasum-core/src/main/kotlin/pw/stamina/occasum/node/visit/AbstractPropertyNodeVisitor.kt
package pw.stamina.occasum.node.visit

import pw.stamina.occasum.node.PropertyNode

abstract class AbstractPropertyNodeVisitor : PropertyNodeVisitor {

    override fun visitChildNode(childNode: PropertyNode): PropertyNodeVisitor? {
        return createChildVisitor(childNode)
    }

    protected abstract fun createChildVisitor(childNode: PropertyNode): PropertyNodeVisitor?
}
